plugins/processors/parser/parser.go

```diff
@@ -16,17 +17,17 @@ type Parser struct {
 	parser       telegraf.Parser
 }
 
-func (p *Parser) Apply(metrics ...telegraf.Metric) []telegraf.Metric {
-	if p.parser == nil {
-		var err error
-		p.parser, err = parsers.NewParser(&p.Config)
-		if err != nil {
-			p.Log.Errorf("could not create parser: %v", err)
-			return metrics
-		}
-		models.SetLoggerOnPlugin(p.parser, p.Log)
+func (p *Parser) Init() (err error) {
+	// Initialise parser
+	p.parser, err = parsers.NewParser(&p.Config)
+	if err != nil {
+		return fmt.Errorf("could not create parser: %v", err)
 	}
 
+	return err
+}
```

Comment:
Before this change, if parsers.NewParser() failed, metrics would pass through and telegraf would continue running. After the change, Init() fails, which causes telegraf to exit. This is a big behavior change.

I think failing Init() and exiting telegraf makes sense if the parser failed to init, but since this isn't how it worked, users will see this as a compatibility change if they do have some parser creation error.

We could change this code to switch to initializing in Init() and not Apply() but we probably should do it in a way that doesn't cause telegraf to exit if there's a failure.

Reply:
🤔 others could say telegraf is now ‘silently’ ignoring wrongly configured parser.

I’m not sure if `p.Log` would already be available during init?

Reply:
@reimda see it as the change that telegraf now halts on wrongly configured duration settings.

Reply:
@reimda ?

Reply:
What is an example of something this would catch?

Reply:
@reimda what would you consider a valid use-case of the current behavior? Why would you want to use misconfigured parser **processor** just to pass through data?

We did not document/advertise this behavior anywhere and I have a really hard time coming up with a valid use case besides from "oh this processor is there unintentionally but it didn't do any harm"... Users can also "fix"/workaround this change by just removing the processor in the config which would restore the previous behavior.

So my take is that this is a good and valid simplification. If you really want, we can just log in `Init` and not error out and add
```golang
if p.parser == nil {
    return metrics
}
```
to the `Apply` function.

Reply:
> What is an example of something this would catch?

NewParser fails on json being given an invalid filter (glob), the graphite or dropwizard parsers being given invalid templates, collectd getting invalid types_db, grok parser getting invalid patterns. I agree all these things are valid errors that should cause telegraf to print an error and exit. The problem is telegraf never has and we don't know how many of our users expect it to work the way it works or have gotten used to it working the way it works.

> "oh this processor is there unintentionally but it didn't do any harm"

A while back I made a change similar to this in the disk input. It was silently ignoring some errors and I while I was cleaning up some other things I changed it to print errors https://github.com/influxdata/telegraf/pull/10527/files#r792929816. It got the attention of lots of people (see 10897) and didn't seem to produce any benefit. Eventually we reverted the code.

This situation with making parser initialization more strict is very similar to the disk input error thing. It's more "correct" to do what Thomas wants, but there's no clear benefit to make a change, no bug report asking for it to be changed, and definite risk that the change will stop telegraf from running for users who have an insignificant problem in their config.

Reply:
If that is what @reimda wants, I will apply the suggestions.. But I don't agree to it.

If users currently configured a not working processor, they have a wrong configuration file.
We had the same with incorrect set durations. An earlier telegraf version would also ignore these values silently and run telegraf without taking care of the intended duration settings. After we fixed it, telegraf would now halt if it encounters such wrongly set durations. IMHO this is the same scenario we have here and therefore telegraf should also halt because of incorrect configuration file.